// yandex-dialogs-sdk/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const alice_1 = __importDefault(require("./alice"));
exports.Alice = alice_1.default;
const scene_1 = __importDefault(require("./scene"));
exports.Scene = scene_1.default;
// fp
const reply_1 = __importDefault(require("./reply"));
exports.reply = reply_1.default;
const button_1 = __importDefault(require("./button"));
exports.button = button_1.default;
// for java-lovers
const replyBuilder_1 = __importDefault(require("./replyBuilder"));
exports.ReplyBuilder = replyBuilder_1.default;
const buttonBuilder_1 = __importDefault(require("./buttonBuilder"));
exports.ButtonBuilder = buttonBuilder_1.default;
// const button = require('./button')
// const reply = require('./reply')
const loggingMiddleware_1 = __importDefault(require("./middlewares/loggingMiddleware"));
exports.loggingMiddleware = loggingMiddleware_1.default;
exports.default = alice_1.default;
module.exports = alice_1.default;
module.exports.Scene = scene_1.default;
module.exports.button = button_1.default;
module.exports.reply = reply_1.default;
module.exports.ButtonBuilder = buttonBuilder_1.default;
module.exports.ReplyBuilder = replyBuilder_1.default;
module.exports.loggingMiddleware = loggingMiddleware_1.default;
