// yandex-dialogs-sdk/dist/logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const colors_1 = __importDefault(require("colors"));
const eventEmitter_1 = __importDefault(require("./eventEmitter"));
const constants_1 = require("./constants");
colors_1.default.setTheme({
    silly: 'rainbow',
    input: 'grey',
    verbose: 'cyan',
    prompt: 'grey',
    info: 'green',
    data: 'grey',
    help: 'cyan',
    warn: 'yellow',
    debug: 'blue',
    error: 'red',
});
const logTime = (timestamp) => {
    const event = new Date(timestamp);
    return event.toLocaleString('en-US');
};
class Logger {
    constructor() {
        eventEmitter_1.default.subscribe(constants_1.EVENT_MESSAGE_RECIEVED, this.log);
        eventEmitter_1.default.subscribe(constants_1.EVENT_MESSAGE_SENT, this.log);
        eventEmitter_1.default.subscribe(constants_1.EVENT_MESSAGE_NOT_SENT, this.log);
    }
    log(event) {
        // tslint:disable-next-line:no-console
        console.log([
            colors_1.default.info(`[${logTime(event.timestamp)}]:`),
            colors_1.default.verbose(`(${event.type})`),
            event.data ? colors_1.default.warn(event.data) : null,
        ].filter(Boolean).join(' '));
    }
}
exports.default = Logger;
