// yandex-dialogs-sdk/dist/tests/reply.spec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const reply_1 = __importDefault(require("../reply"));
const card_1 = require("../card");
const constants_1 = require("../constants");
test('create reply with string constructor', () => {
    const expectedData = {
        response: {
            text: 'send message',
            buttons: [],
            end_session: constants_1.DEFAULT_END_SESSION,
        },
        session: null,
        version: constants_1.ALICE_PROTOCOL_VERSION,
    };
    const msg = reply_1.default(expectedData.response.text);
    expect(msg).toEqual(expectedData);
});
test('create reply with object constructor', () => {
    const expectedData = {
        response: {
            text: 'send message',
            tts: 'send m+essage',
            buttons: [],
            end_session: true,
        },
        session: null,
        version: constants_1.ALICE_PROTOCOL_VERSION,
    };
    const msg = reply_1.default({
        text: expectedData.response.text,
        tts: expectedData.response.tts,
        buttons: expectedData.response.buttons,
        endSession: expectedData.response.end_session,
    });
    expect(msg).toEqual(expectedData);
});
test('creating big image card', () => {
    const mockedImageId = '213044/2a175da14f91b71df60c';
    expect(card_1.image(mockedImageId)).toEqual({
        image_id: mockedImageId,
    });
});
